File: tree.py
import math
import numpy
import random
import scipy.special


class Tree:
    def __init__(self, sample_size):
        self.left_child = numpy.full(2 * sample_size - 1, -1)
        self.right_child = numpy.full(2 * sample_size - 1, -1)
        self.parent = numpy.full(2 * sample_size - 1, -1)
        self.time = numpy.zeros(2 * sample_size - 1)
        self.sample_size = sample_size

        t = 0
        active_lineages = list(range(sample_size))
        n = len(active_lineages)
        next_parent = sample_size
        while n > 1:
            t += random.expovariate(lambd=scipy.special.binom(n, 2))
            [l_child, r_child] = numpy.random.choice(
                active_lineages, size=2, replace=False
            )
            self.parent[l_child] = next_parent
            self.parent[r_child] = next_parent
            self.left_child[next_parent] = l_child
            self.right_child[next_parent] = r_child
            self.time[next_parent] = t
            active_lineages.remove(l_child)
            active_lineages.remove(r_child)
            active_lineages.append(next_parent)
            next_parent += 1
            n -= 1

    def sample_reattach_time(self, child, new_sib):
        sib = self.sibling(child)
        new_sib_parent = self.parent[new_sib]
        new_sib_grandparent = self.grandparent(new_sib)
        if sib == new_sib:
            if new_sib_grandparent == -1:
                ret = self.time[new_sib] + random.expovariate(lambd=1)
            else:
                ret = numpy.random.uniform(
                    self.time[new_sib], self.time[new_sib_grandparent]
                )
        else:
            if new_sib_parent == -1:
                ret = self.time[new_sib] + random.expovariate(lambd=1)
            else:
                ret = numpy.random.uniform(
                    self.time[new_sib], self.time[new_sib_parent]
                )
        return ret

    def log_reattach_density(self, child, new_sib, new_time):
        sib = self.sibling(child)
        new_sib_parent = self.parent[new_sib]
        new_sib_grandparent = self.grandparent(new_sib)
        if sib == new_sib:
            if new_sib_grandparent == -1:
                ret = self.time[new_sib] - new_time
            else:
                ret = -math.log(self.time[new_sib_grandparent] - self.time[new_sib])
        else:
            if new_sib_parent == -1:
                ret = self.time[new_sib] - new_time
            else:
                ret = -math.log(self.time[new_sib_parent] - self.time[new_sib])
        return ret

    def resample_times(self):
        ret = 0
        lb = 0
        [sorted_times, ind] = numpy.unique(self.time, return_index=True)
        for i in range(1, len(sorted_times)):
            rate = scipy.special.binom(self.sample_size - i + 1, 2)
            index = ind[i]
            self.time[index] = lb + random.expovariate(lambd=rate)
            ret -= rate * (self.time[index] - lb)
            lb = self.time[index]
        return ret

    def log_resample_times_density(self, t):
        ret = 0
        lb = 0
        [sorted_times, ind] = numpy.unique(t, return_index=True)
        for i in range(1, len(sorted_times)):
            rate = scipy.special.binom(self.sample_size - i + 1, 2)
            index = ind[i]
            ret -= rate * (t[index] - lb)
            lb = t[index]
        return ret

    def sample_leaf(self):
        node = numpy.random.choice(numpy.arange(self.sample_size))
        return node

    def sample_node(self):
        node = numpy.random.choice(len(self.parent))
        return node

    def replace_child(self, node, child, new_child):
        if self.left_child[node] == child:
            self.left_child[node] = new_child
        else:
            self.right_child[node] = new_child

    def detach_reattach(self, child, new_sib, new_time):
        sib = self.sibling(child)
        child_parent = self.parent[child]
        sib_grandparent = self.grandparent(child)
        new_sib_parent = self.parent[new_sib]
        self.time[child_parent] = new_time
        if sib != new_sib:
            self.parent[new_sib] = child_parent
            self.parent[sib] = sib_grandparent
            self.parent[child_parent] = new_sib_parent
            self.replace_child(child_parent, sib, new_sib)
            if sib_grandparent != -1:
                self.replace_child(sib_grandparent, child_parent, sib)
            if new_sib_parent != -1:
                self.replace_child(new_sib_parent, new_sib, child_parent)

    def sibling(self, node):
        ret = -1
        p = self.parent[node]
        if p != -1:
            ret = self.left_child[p]
            if ret == node:
                ret = self.right_child[p]
        return ret

    def grandparent(self, node):
        ret = -1
        if self.parent[node] != -1:
            ret = self.parent[self.parent[node]]
        return ret

    def log_likelihood(self):
        sorted_times = numpy.unique(self.time)
        ret = 0
        for i in range(self.sample_size - 1):
            ret -= scipy.special.binom(self.sample_size - i, 2) * (
                sorted_times[i + 1] - sorted_times[i]
            )
        return ret

File: C/tree.c
#include "tree.h"
#include <gsl/gsl_cdf.h>
#include <gsl/gsl_randist.h>
#include <gsl/gsl_rng.h>
#include <math.h>

int compare(const void *a, const void *b) {
  int ret = 0;
  if (*(double *)a > *(double *)b) {
    ret = 1;
  } else if (*(double *)a < *(double *)b) {
    ret = -1;
  }
  return ret;
}

int compare_with_index(const void *a, const void *b) {
  int ret = 0;
  sortee *a1 = (sortee *)a;
  sortee *b1 = (sortee *)b;
  if (a1->value > b1->value) {
    ret = 1;
  } else if (a1->value < b1->value) {
    ret = -1;
  }
  return ret;
}

void init_tree(tree *t, const int n) {
  t->left_child = malloc((2 * n - 1) * sizeof(int));
  t->right_child = malloc((2 * n - 1) * sizeof(int));
  t->parent = malloc((2 * n - 1) * sizeof(int));
  t->time = malloc((2 * n - 1) * sizeof(double));
  t->sample_size = n;
  for (int i = 0; i < n; i++) {
    t->left_child[i] = -1;
    t->right_child[i] = -1;
    t->time[i] = 0.0;
  }
  t->parent[2 * n - 2] = -1;
  return;
}

void generate_tree(tree *t, gsl_rng *gen) {
  double r = 0;
  int n = t->sample_size;
  int active_lineages[n];
  for (int i = 0; i < n; i++) {
    active_lineages[i] = i;
  }
  int next_parent = n;
  int l_child, r_child;
  while (n > 1) {
    r += gsl_ran_exponential(gen, 2.0 / (n * (n - 1)));
    l_child = floor(n * gsl_rng_uniform(gen));
    do {
      r_child = floor(n * gsl_rng_uniform(gen));
    } while (l_child == r_child);
    t->parent[active_lineages[l_child]] = next_parent;
    t->parent[active_lineages[r_child]] = next_parent;
    t->left_child[next_parent] = active_lineages[l_child];
    t->right_child[next_parent] = active_lineages[r_child];
    t->time[next_parent] = r;
    active_lineages[(int)fmin(l_child, r_child)] = next_parent;
    for (int i = (int)fmax(l_child, r_child); i < n - 1; i++) {
      active_lineages[i] = active_lineages[i + 1];
    }
    next_parent++;
    n--;
  }
  return;
}

void print_tree(tree *t) {
  for (int i = 0; i < 2 * t->sample_size - 1; i++) {
    printf("%d, %d, %d, %d, %f\n", i, t->parent[i], t->left_child[i],
           t->right_child[i], t->time[i]);
  }
  return;
}

int sibling(tree *t, const int node) {
  int ret = -1;
  int p = t->parent[node];
  if (p != -1) {
    ret = t->left_child[p];
    if (ret == node) {
      ret = t->right_child[p];
    }
  }
  return ret;
}

int grandparent(tree *t, const int node) {
  int ret = -1;
  int p = t->parent[node];
  if (p != -1) {
    ret = t->parent[p];
  }
  return ret;
}

double sample_reattach_time(tree *t, const int child, const int new_sib,
                            gsl_rng *gen) {
  int sib = sibling(t, child);
  int new_sib_parent = t->parent[new_sib];
  int new_sib_grandparent = grandparent(t, new_sib);
  double ret = 0.0;
  if (sib == new_sib) {
    if (new_sib_grandparent == -1) {
      ret = t->time[new_sib] + gsl_ran_exponential(gen, 1.0);
    } else {
      ret =
          t->time[new_sib] + (t->time[new_sib_grandparent] - t->time[new_sib]) *
                                 gsl_rng_uniform(gen);
    }
  } else {
    if (new_sib_parent == -1) {
      ret = t->time[new_sib] + gsl_ran_exponential(gen, 1.0);
    } else {
      ret = t->time[new_sib] +
            (t->time[new_sib_parent] - t->time[new_sib]) * gsl_rng_uniform(gen);
    }
  }
  return ret;
}

double log_reattach_density(tree *t, const int child, const int new_sib,
                            const double new_time) {
  double ret = 0.0;
  int sib = sibling(t, child);
  int new_sib_parent = t->parent[new_sib];
  int new_sib_grandparent = grandparent(t, new_sib);
  if (sib == new_sib) {
    if (new_sib_grandparent == -1) {
      ret = t->time[new_sib] - new_time;
    } else {
      ret = -log(t->time[new_sib_grandparent] - t->time[new_sib]);
    }
  } else {
    if (new_sib_parent == -1) {
      ret = t->time[new_sib] - new_time;
    } else {
      ret = -log(t->time[new_sib_parent] - t->time[new_sib]);
    }
  }
  return ret;
}

double resample_times(tree *t, gsl_rng *gen) {
  double ret = 0.0;
  double lb = 0.0;
  int n = t->sample_size;
  sortee sorted_times[n - 1];
  for (int i = 0; i < n - 1; i++) {
    sorted_times[i].value = t->time[n + i];
    sorted_times[i].index = i;
  }
  qsort(sorted_times, n - 1, sizeof(sortee), compare_with_index);
  double rate = 0.0;
  for (int i = 0; i < n - 1; i++) {
    rate = (n - i) * (n - i - 1) / 2.0;
    t->time[n + sorted_times[i].index] =
        lb + gsl_ran_exponential(gen, 1.0 / rate);
    ret -= rate * (t->time[n + sorted_times[i].index] - lb);
    lb = t->time[n + sorted_times[i].index];
  }
  return ret;
}

double log_resample_times_density(tree *t, const double *r) {
  double ret = 0.0;
  int n = t->sample_size;
  double sorted_times[n];
  for (int i = 0; i < n; i++) {
    sorted_times[i] = r[n + i - 1];
  }
  qsort(sorted_times, n, sizeof(double), compare);
  for (int i = 0; i < n - 1; i++) {
    ret -=
        (sorted_times[i + 1] - sorted_times[i]) * (n - i) * (n - i - 1) / 2.0;
  }
  return ret;
}

int sample_leaf(tree *t, gsl_rng *gen) {
  int n = t->sample_size;
  int ret = floor(n * gsl_rng_uniform(gen));
  return ret;
}

int sample_node(tree *t, gsl_rng *gen) {
  int n = t->sample_size;
  int ret = floor((2 * n - 1) * gsl_rng_uniform(gen));
  return ret;
}

void replace_child(tree *t, const int node, const int child,
                   const int new_child) {
  if (t->left_child[node] == child) {
    t->left_child[node] = new_child;
  } else {
    t->right_child[node] = new_child;
  }
  return;
}

void detach_reattach(tree *t, const int child, const int new_sib,
                     const double new_time) {
  int sib = sibling(t, child);
  int child_parent = t->parent[child];
  int sib_grandparent = grandparent(t, child);
  int new_sib_parent = t->parent[new_sib];
  t->time[child_parent] = new_time;
  if (sib != new_sib) {
    t->parent[new_sib] = child_parent;
    t->parent[sib] = sib_grandparent;
    t->parent[child_parent] = new_sib_parent;
    replace_child(t, child_parent, sib, new_sib);
    if (sib_grandparent != -1) {
      replace_child(t, sib_grandparent, child_parent, sib);
    }
    if (new_sib_parent != -1) {
      replace_child(t, new_sib_parent, new_sib, child_parent);
    }
  }
  return;
}

double log_likelihood(tree *t) {
  double ret = 0.0;
  int n = t->sample_size;
  double sorted_times[n];
  for (int i = 0; i < n; i++) {
    sorted_times[i] = t->time[n + i - 1];
  }
  qsort(sorted_times, n, sizeof(double), compare);
  for (int i = 0; i < n - 1; i++) {
    ret -=
        (n - i) * (n - i - 1) * (sorted_times[i + 1] - sorted_times[i]) / 2.0;
  }
  return ret;
}

void free_tree(tree *t) {
  free(t->left_child);
  free(t->right_child);
  free(t->parent);
  free(t->time);
  free(t);
  return;
}
